io/include/pcl/io/real_sense_2_grabber.h

```diff
@@ -0,0 +1,176 @@
+ *
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of the copyright holder(s) nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#pragma once
+
+#include <pcl/io/boost.h>
+#include <pcl/io/grabber.h>
+
+#include <librealsense2/rs.hpp>
+
+namespace pcl
+{
+  struct PointXYZ;
+  struct PointXYZRGB;
+  struct PointXYZRGBA;
+  struct PointXYZI;
+  template <typename T> class PointCloud;
+
+  /** \brief Grabber for Intel Realsense 2 SDK devices (D400 series)
+  * \author Patrick Abadi <patrickabadi@gmail.com>, Daniel Packard <pack3754@gmail.com>
+  * \ingroup io
+  */
+  class PCL_EXPORTS RealSense2Grabber : public pcl::Grabber
+  {
+  public:
+    /** \brief Constructor
+    * \param[in] serial_number optional to start up a specific device
+    */
+    RealSense2Grabber (const uint32_t serial_number = 0);
+
+    /** \brief Constructor
+    * \param[in] file_name used for loading sample data bag file
+    */
+    RealSense2Grabber (const std::string& file_name);
+
+    /** \brief virtual Destructor inherited from the Grabber interface. It never throws. */
+    virtual ~RealSense2Grabber () throw ();
+
+    /** \brief Set the device options
+    * \param[in] width resolution
+    * \param[in] height resolution
+    * \param[in] fps target frames per second for the device
+    */
+    inline void 
+    setDeviceOptions (uint32_t width, uint32_t height, uint32_t fps = 30)
+    {
+      device_width_ = width;
+      device_height_ = height;
+      target_fps_ = fps;
+    }
```

Comment:
I have a consultation about the specification.
Currently, This feature is only valid before <code>RealSense2Grabber::start()</code>.
Do you think that this feature should be enabled even after <code>RealSense2Grabber::start()</code>?

```patch
    inline void 
    setDeviceOptions (uint32_t width, uint32_t height, uint32_t fps = 30)
    {
+     std::lock_guard<std::mutex> guard (mutex_);
+
      device_width_ = width;
      device_height_ = height;
      target_fps_ = fps;
+
+     try
+     {
+       pipe_.stop();
+     }
+     catch (const rs2_extension& ignored)
+     {
+     }
+    
+     rs2::config cfg;
+     cfg.enable_stream (RS2_STREAM_COLOR, device_width_, device_height_, RS2_FORMAT_RGB8, target_fps_);
+     cfg.enable_stream (RS2_STREAM_DEPTH, device_width_, device_height_, RS2_FORMAT_ANY, target_fps_);
+     cfg.enable_stream (RS2_STREAM_INFRARED, device_width_, device_height_, RS2_FORMAT_ANY, target_fps_);
+    
+     pipe_.start (cfg);
    }
```

Reply:
I agree that this should be handled. Do you want to 
- A. Print an error informing the user that these settings can't be updated while the sensor is running, and return early before updating the member variables
- or B. Print a warning informing the user that changes will not be applied until the sensor is stopped/started, and set the member variables so they will be picked up after the next stop/start
- or C. ... other?

Reply:
I updated setDeviceOptions so it would restart if it was currently running.